pkg/api/metrics.go

```diff
@@ -31,42 +32,23 @@ func (hs *HTTPServer) QueryMetricsV2(c *models.ReqContext, reqDTO dtos.MetricReq
 		Queries:   make([]plugins.DataSubQuery, 0, len(reqDTO.Queries)),
 	}
 
-	// Loop to see if we have an expression.
-	prevType := ""
-	var ds *models.DataSource
+	// Parse the queries
+	hasExpression := false
+	datasources := make(map[string]*models.DataSource, len(reqDTO.Queries))
 	for _, query := range reqDTO.Queries {
-		dsType := query.Get("datasource").MustString("")
-		if dsType == expr.DatasourceName {
-			return hs.handleExpressions(c, reqDTO)
+		ds, errRsp := hs.getDataSourceFromQuery(c, query, datasources)
+		if errRsp != nil {
+			return errRsp
 		}
-		if prevType != "" && prevType != dsType {
-			// For mixed datasource case, each data source is sent in a single request.
-			// So only the datasource from the first query is needed. As all requests
-			// should be the same data source.
-			hs.log.Debug("Can't process query since it's missing data source ID")
-			return response.Error(http.StatusBadRequest, "All queries must use the same datasource", nil)
+		if ds == nil {
+			return response.Error(http.StatusBadRequest, "Datasource not found for query", nil)
 		}
 
-		if ds == nil {
-			// require ID for everything
-			dsID, err := query.Get("datasourceId").Int64()
-			if err != nil {
-				hs.log.Debug("Can't process query since it's missing data source ID")
-				return response.Error(http.StatusBadRequest, "Query missing data source ID", nil)
-			}
-			if dsID == grafanads.DatasourceID {
-				ds = grafanads.DataSourceModel(c.OrgId)
-			} else {
-				ds, err = hs.DataSourceCache.GetDatasource(dsID, c.SignedInUser, c.SkipCache)
-				if err != nil {
-					return hs.handleGetDataSourceError(err, dsID)
-				}
-			}
+		datasources[ds.Uid] = ds
+		if ds.Uid == expr.DatasourceUID {
```

Comment:
Based on earlier suggestion in `getDataSourceFromQuery`.

```suggestion
		if expr.IsDataSource(ds.Uid) {
```